org.scala-ide.sdt.core.tests/src/org/scalaide/core/semantichighlighting/classifier/RegionParser.scala

```diff
@@ -2,10 +2,68 @@ package org.scalaide.core.semantichighlighting.classifier
 
 import org.eclipse.jface.text.IRegion
 import org.eclipse.jface.text.Region
+import scala.annotation.tailrec
+
+import language.implicitConversions
 
 object RegionParser {
 
   /**
+   * This class represents a substring with an optional prefix and suffix.
+   */
+  final case class EmbeddedSubstr(str: String, prefix: String = "", suffix: String = "") {
+    private[RegionParser] val searchString = prefix + str + suffix
+  }
+
+  object EmbeddedSubstr {
+    implicit def wrapAsEmbeddedSubstring(str: String) = EmbeddedSubstr(str)
+  }
+
+
+  /**
+   * Extracts the regions marked by the given substrings.
+   *
+   * @example
+   * In the easiest case this function just extracts substrings (note that there is an implicit conversion
+   * from [[[String]]] to [[[org.scalaide.core.semantichighlighting.classifier.RegionParser.EmbeddedSubstr]]]):
+   * {{{
+   * scala> RegionParser.substrRegions("x + f(x)", "x")
+   * res0: Map(offset: 0, length: 1 -> EmbeddedSubstr(x,,), offset: 6, length: 1 -> EmbeddedSubstr(x,,))
+   * }}}
+   *
+   * @example
+   * Extract occurrences of `x` when surrounded by parenthesis:
+   * {{{
+   *  scala> substrRegions("x + f(x)", EmbeddedSubstr("x", "(", ")"))
+   *  res0: Map(offset: 6, length: 1 -> EmbeddedSubstr(x,(,)))
+   * }}}
+   *
+   * @example
+   * Here we extract both occurrences of `x` separately:
+   * {{{
+   * scala> RegionParser.substrRegions("x + f(x)", EmbeddedSubstr("x", suffix = " "), EmbeddedSubstr("x", "(", ")"))
+   * res0: Map(offset: 0, length: 1 -> EmbeddedSubstr(x,, ), offset: 6, length: 1 -> EmbeddedSubstr(x,(,)))
+   * }}}
+   */
+  def substrRegions(test: String, substrs: EmbeddedSubstr*): Map[IRegion, EmbeddedSubstr] = {
```

Comment:
I think a solution similar to `delimitedRegions` would result in a better test suite. You can just surround a block by a matching pair of signs:
```scala
f([5])
def f(param: => Int) = param
```
Even nesting works:
```scala
f([f([5])])
def f(param: => Int) = param
```
You are just not allowed to use the pair signs anymore, but that shouldn't be a problem.